### DISCUSSION TITLE
Failed to connect to ESP32: No serial data received. (EVTV ESP32 CANdue)

### DISCUSSION THREAD

**Discussion by JoeBargo** [collin80/SavvyCAN#573] [Q&A]
I couldn't get ESP32RET to load onto my device (verified fine and changed partition scheme). I keep getting "exit status 2". I tried just loading an example from your esp32_can examples and same issue. 

I see 4 NEW ports when I connect (on a mac by the way):
-USBSerial-1400 
-USBSerial-1401
-COM 5
-COM 6

I tried a few combinations of holding boot and reset on the board, but I'm not sure if that's the issue..

Any guidance would help,
Thanks

**Comment by JoeBargo** [comment]
> I couldn't get ESP32RET to load onto my device (verified fine and changed partition scheme). I keep getting "exit status 2". I tried just loading an example from your esp32_can examples and same issue. 
> 
> 
> 
> I see 4 NEW ports when I connect (on a mac by the way):
> 
> -USBSerial-1400 
> 
> -USBSerial-1401
> 
> -COM 5
> 
> -COM 6
> 
> 
> 
> I tried a few combinations of holding boot and reset on the board, but I'm not sure if that's the issue..
> 
> 
> 
> Any guidance would help,
> 
> Thanks

Also, serial monitor just shows random characters at all baud rates. 

GVRET Updater on the savvycan website isn't able to find the device. ESP32RET updater either gives the same error (no serial data received) or "timed out waiting for packet header".

I tried holding boot, pressing reset, the releasing boot at various lengths and points in flashing.

**Comment by collin80** [comment]
You're getting COM ports on a Mac?! Are you apparently running Windows with Bootcamp? ESP32RET uses 1000000 baud which may be considered somewhat uncommon. So, one thing to to ensure you've tried that speed. 

The GVRET updater is for the older SAM3X processor based boards. So, it is correct that it cannot find any boards to flash. 

On the ESP32 you will want to use the first serial port it creates. But, it should only create two, not four.

If you are using MacOS, which version of MacOS do you have?

**Reply by JoeBargo** [reply]
I think the 4 ports are just 2 ports. For example, if USBSerial-5 is actively monitored, USBSerial-1400 will be busy. Same with -6 and -1401. I'm not sure what that means or if it might help explain why I'm having trouble flashing.

**Comment by JoeBargo** [comment]
> You're getting COM ports on a Mac?! Are you apparently running Windows with Bootcamp? ESP32RET uses 1000000 baud which may be considered somewhat uncommon. So, one thing to to ensure you've tried that speed. 
> 
> 
> 
> The GVRET updater is for the older SAM3X processor based boards. So, it is correct that it cannot find any boards to flash. 
> 
> 
> 
> On the ESP32 you will want to use the first serial port it creates. But, it should only create two, not four.
> 
> 
> 
> If you are using MacOS, which version of MacOS do you have?


I’ve been using MacOS Ventura 13.x with an M1 chip (MacBook) and Win10 on separate computer. I have parallels too for the mac but I’ve stopped messing with that since I didn’t want another potential issue to come up. I’m switching back and forth between a windows pc and mac trying a lot of stuff. I mixed up the port names but both create 4 for some reason. 

So on mac it’s naming them something like this USBSerial-17, USBSerial-18, USBSerial-100, USBSerial-101. Every time I unplug and replug the device in it changes the name to USBSerial-19, USBSerial-20 (on Mac). Normally the other 2 don’t change, but every few attempts it changes to USBSerial-1000, USBSerial-1001. 

On windows or macOS I can’t get it to upload anything. However, I was able to use a web flasher and manually added the partitions (like ESP32RET’s batch/command does). Ive tried esptool on both Mac and windows but both fail to upload. 

Using the esp32 web flasher I uploaded an example from your esp32-can repository. I did this using the compiled binaries that arduino IDE made. I can now see what I’m supposed to be seeing for 1 of the connections, but the other is still gibberish. Maybe it’s because CAN FD’s default is 4000000? I’m going to recompile ESP32RET later with that changed to 500000 and see if that works.

So when I do this, how do I get the program to work use GVRET format in savvycan? Do I need to do something to tell the device that’s what output I want or does it just know? Connecting it out of the box, I created a serial GVRET (500000) and nothing happened. Then I tried adding a lawicel connection (112500,500000) and got a ton of useless hex in the console. I think that was the gibberish I was seeing in the serial monitor on the incorrect port.

I don't know why Im having issues flashing but I’ve reinstalled the ftdi vcp drivers, cp2102 drivers and ch340 drivers. I think it just needs the ftdi vcp drivers but I had the cp2102 drivers from another board.

**Comment by JoeBargo** [comment]
<img width="1411" alt="image" src="https://user-images.githubusercontent.com/11463120/216779323-8a3c4a6c-af2e-44ba-b467-15a60cc238a1.png">

So I think I got it working. Still a little confused why the flashing is being difficult, but getting 3000+ frames/sec is worth it.

**Comment by collin80** [comment]
Yes, I don't know why flashing was so difficult. But, I'm glad it sounds like it's working. Now that you mention it, I suppose on MacOS you do kind of get 4 ports. They're really two ports but each presented two different ways. I have no idea why MacOS does that. Usually they're labeled CU and TTY and usually the tty version is the one to use. 

It's strange it took so long to flash properly but the good news is that you should only have to do that once.

**Reply by JoeBargo** [reply]
Yeah all 4 have .cu and .tty. I think it has a port for the serial connection and the converter? No idea.

In another post, you said something about being able to start the device using USB and connecting via Wi-Fi. How are you able to keep it initialized after unplugging the USB using the 12V power? 

Thanks for the help!

**Reply by collin80** [reply]
Yes, the hardware is wifi capable so if you power it via 12v you can leave it connected to the CAN bus and connect via WiFi to get CAN captures and/or send traffic. It works wirelessly to connect to SavvyCAN. By default I believe it does have WiFi turned on. You will probably see an SSID called ESP32SSID or something like that. If you connect your machine to that wifi then you can use "Network Connection (GVRET)" from SavvyCAN. If you're connected to wifi you should automatically see the IP address of the ESP32 show up in the list. You can change the SSID or make the ESP32 board connect to your existing SSID which tends to be about 1000x nicer since then you still can have internet. To do that you currently need to access the serial console. If you have a serial program like CoolTerm then you can point it at the serial port for the ESP32 that you connect to SavvyCAN with. The speed is 1 megabit. Then have some sort of line ending and type ? to get a menu. The commands to set an SSID, password, and tell it to connect are as follows:
SSID=yourSSID
WPA2KEY=yourWPA2password
WIFIMODE=1

Thereafter it will be able to connect to your existing wireless. This is great if you have it in one place and you know you'll be in wifi range. Otherwise the default of it making its own SSID tends to be better.

**Reply by JoeBargo** [reply]
Yeah I've been using it's as WIFIMODE 2 since I can't actually get serial GVRET to work. I'll try using my phones hotspot or something for WIFIMODE 1.

Is there any way to initialize wifi during the boot cycle? (Or any way where I don't need to plug in to USB?

Also, I switched to the most recent build and fixed some settings that were an issue for me before. Is there a way to output a GVRET file for each data log? You had something like this in the ESP32RET Updater build but I don't think it actually made a file (or I couldn't find it). I would use SavvyCan but it tends to crash after a couple minutes of logging. 

~~Also, are the filters just commented out in the build or do they not work?~~ <- disregard, I got filters working looking at some of your other stuff.

Thanks for the support again
